src/qml/qmlwaveformrenderer.cpp

```diff
@@ -0,0 +1,231 @@
+#include "qml/qmlwaveformrenderer.h"
+
+#include "moc_qmlwaveformrenderer.cpp"
+#include "util/assert.h"
+#include "waveform/renderers/allshader/waveformrenderbeat.h"
+#include "waveform/renderers/allshader/waveformrendererendoftrack.h"
+#include "waveform/renderers/allshader/waveformrendererpreroll.h"
+#include "waveform/renderers/allshader/waveformrendererrgb.h"
+#ifdef __STEM__
+#include "waveform/renderers/allshader/waveformrendererstem.h"
+#endif
+#include "waveform/renderers/allshader/waveformrendermark.h"
+#include "waveform/renderers/allshader/waveformrendermarkrange.h"
+
+namespace mixxx {
+namespace qml {
+
+QmlWaveformRendererMark::QmlWaveformRendererMark()
+        : m_defaultMark(nullptr),
+          m_untilMark(std::make_unique<QmlWaveformUntilMark>()) {
+}
+
+QmlWaveformRendererFactory::Renderer QmlWaveformRendererEndOfTrack::create(
+        WaveformWidgetRenderer* waveformWidget) const {
+    auto* pRenderer = new allshader::WaveformRendererEndOfTrack(waveformWidget);
```

Comment:
`parented_ptr`?

Reply:
Unfortunately, `WaveformWidgetRenderer` doesn't inherit `QObject` so `parented_ptr` cannot be used. Note that the ptr is eventually encapsulated inside a `unique_ptr` [here](https://github.com/mixxxdj/mixxx/blob/5a7e04b80ee6a1c8c54f06ba37193fd975123171/src/qml/qmlwaveformdisplay.cpp#L129) to ensure proper de-allocation.

Reply:
Oh, I just assumed it was a QObject and thus managed by the QObject tree. So this is actually a raw owning pointer? In that case we should definitely work a little on this. You can easily convert a `std::unique_ptr<Derived>` to a `std::unique_ptr<Base>` (the types are just not covariant compared to raw pointers). I'm fairly certain you can `std::make_unique` from the start and then also move that into the `QmlWaveformRendererFactory::Renderer`.

Reply:
No, unfortunately I can't because I have to pass that pointer twice inside the Render struct. This is because I need to keep track of the two base classes of the renderers (WaveformRenderer and Rendergraph node, unrelated together). I'm sure this is very hacky but unfortunately I cannot think of any other ways to wrap that elegantly, without adding some bloating on allshader and rendergraph.
Happy to try an alternative if you have ideas. Note that I cannot use a shared_ptr since one of its usage (inside rendergraph) will be downcasted to unique_ptr

Reply:
I think its fairly straightforward if adapt something like this:

```cpp

class A {};
class B {};
class C : public A, public B {};

struct Renderer {
    // ensure owning ptr is before the borrowing one
    // to ensure the borrow doesn't outlive the owner.
    std::unique_ptr<A> node{nullptr};
    B* renderer{nullptr};
};

Renderer create() {
    auto c = std::make_unique<C>();
    // local temporary is needed so we don't call `.get` on `c` after it may have been moved from
    B* b = c.get();
    return Renderer{std::move(c), b};
}

```

wdyt?

Reply:
Yep, works with me. I thought it was a bit strange to use a mixture of both, but happy to do so if you think it's the best tradeoff

Reply:
I think so, yes. Lets try our best to adhere to [`ES.60: Avoid new and delete outside resource management functions`](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es60-avoid-new-and-delete-outside-resource-management-functions)